buildbot: 0.9.0rc4 -> 0.9.0.post1

nand0p opened the pull request:
buildbot: 0.9.0rc4 -> 0.9.0.post1
- updates buildbot to version 9 release
- adds nixos configuration module
- fixes buildbot-www package deps
- re-hardcode path to tail

Commit:
buildbot: 0.9.0rc4 -> 0.9.0.post1

- updates buildbot to version 9 release
- adds nixos configuration module
- fixes buildbot-www package deps
- re-hardcode path to tail
- builbot configuration via module vars

nand0p commented:
@FRidh @copumpkin @ryansydnor this PR is WIP, as still working thru the buildbot nixos configuration module.

Mic92 commented on nixos/modules/services/continuous-integration/buildbot/default.nix:
```diff
@@ -0,0 +1,111 @@
+      script = "${pkgs.buildbot-ui}/bin/buildbot start master";
+
+      postStart = ''
+        until [[ $(${pkgs.curl}/bin/curl -s --head -w '\n%{http_code}' http://localhost:${toString cfg.port} | tail -n1) =~ ^(200|403)$ ]]; do
```
what is that for?

grahamc replied:
This runs curl until it returns a status code indicating it is running, to block "After"-dependent services.

Mic92 replied:
Do we have any services with an "After" dependency on this?

nand0p replied:
buildbot worker would need buildbot master running to come up successfully

copumpkin replied:
@nand0p would you run a worker on the same machine as a master?

nand0p replied:
in a dev/test environment, bb master and worker may run on same machine. This is probably a common setup.

Unlikely in production, although i can imagine a scenario where a dedicated builder/worker would need to run on master (to make use of instance profile, secrets, etc), but that would be an edge case, and there are probably better ways to set worker environment.

nand0p commented:
@copumpkin @FRidh @grahamc @Mic92 ready_for_review

copumpkin commented:
I haven't looked deeply yet, but two things jump to mind:

1. You need to include your module in the big list o' modules that NixOS includes (https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/module-list.nix)
2. I'd probably call the module `buildbot-master` or something similar, since it seems likely that people will eventually want a `buildbot-worker` module to handle worker configuration.

copumpkin commented on nixos/modules/services/continuous-integration/buildbot-master/default.nix:
```diff
@@ -0,0 +1,127 @@
+      environment =
+        let
+          selectedSessionVars =
+            lib.filterAttrs (n: v: builtins.elem n [ "NIX_PATH" ])
```
`lib` should re-export user-facing `builtins`, so `lib.elem` would probably be nicer here.

copumpkin commented on nixos/modules/services/continuous-integration/buildbot-master/default.nix:
```diff
@@ -0,0 +1,127 @@
+        ${pkgs.buildbot-ui}/bin/buildbot create-master ${cfg.buildbotDir}
+        sed -i '/www/,+1 d' ${cfg.buildbotDir}/master.cfg.sample
+        mv -v ${cfg.buildbotDir}/master.cfg.sample ${cfg.buildbotDir}/master.cfg
+        echo "c['www'] = dict(port=8010)" >> ${cfg.buildbotDir}/master.cfg
```
I think you mean `${cfg.port}` here, right?

copumpkin commented on nixos/modules/services/continuous-integration/buildbot-master/default.nix:
```diff
@@ -0,0 +1,127 @@
+        sed -i '/www/,+1 d' ${cfg.buildbotDir}/master.cfg.sample
+        mv -v ${cfg.buildbotDir}/master.cfg.sample ${cfg.buildbotDir}/master.cfg
+        echo "c['www'] = dict(port=8010)" >> ${cfg.buildbotDir}/master.cfg
+        cat ${cfg.buildbotDir}/master.cfg
```
It seems like it would be nice for master.cfg to exist in the store, to avoid the config from changing without Nix being aware of it (since it'll just overwrite it anyway).

Maybe use `-c` to `create-master` to point it at a master.cfg in the store?

copumpkin commented on nixos/modules/services/continuous-integration/buildbot-master/default.nix:
```diff
@@ -0,0 +1,243 @@
+          selectedSessionVars //
+            { BUILDBOT_HOME = cfg.home;
+              BUILDBOT_DIR = cfg.buildbotDir;
+              NIX_REMOTE = "daemon";
```
Hmm, what's this about?

copumpkin commented on nixos/modules/services/continuous-integration/buildbot-master/default.nix:
```diff
@@ -0,0 +1,243 @@
+  cfg = config.services.buildbot-master;
+  configFile = pkgs.writeText "master.cfg" ''
+
+from buildbot.plugins import *
```
There's no need to not indent this, by the way. The double quote thing looks for the smallest amount of leading whitespace and subtracts that from all lines, IIRC.

copumpkin commented on pkgs/development/tools/build-managers/buildbot/default.nix:
```diff
@@ -57,12 +57,13 @@ pythonPackages.buildPythonApplication (rec {
 
   postFixup = ''
+    # re-hardcode path to tail
+    sed -i 's|/usr/bin/tail|${coreutils}/bin/tail|' $out/lib/python2.7/site-packages/buildbot/scripts/logwatcher.py
```
If you did this in the `patchPhase` or `pre/postPatch`, you wouldn't have to remove the .pyc file, right?

copumpkin commented on nixos/modules/services/continuous-integration/buildbot-master/default.nix:
```diff
@@ -0,0 +1,243 @@
+
+      changeSource = mkOption {
+        default = " \\
+          \"changes.GitPoller('git://github.com/buildbot/pyflakes.git', workdir='gitpoller-workdir', branch='master', pollinterval=300)\" \\
```
These defaults seem more appropriate for `example`, than `default`. `example` just shows up in the documentation as possible values, whereas `default` will actually get set if you don't set some other value for it. Given that most people probably don't need to monitor the pyflakes repo, it makes more sense to me for the config to complain that this hasn't been set than to default to looking somewhere many users might not want to look.

copumpkin commented on nixos/modules/module-list.nix:
```diff
@@ -134,6 +134,7 @@
   ./services/computing/torque/server.nix
   ./services/computing/torque/mom.nix
   ./services/computing/slurm/slurm.nix
+  ./services/continuous-integration/buildbot-master/default.nix
```
I'd probably either do ./services/continuous-integration/buildbot/master.nix (and then someday we might get a /worker.nix) or buildbot-master.nix. Most services don't use a default.nix or have supporting files.

nand0p commented:
ready_for_review

grahamc commented:
Bump @copumpkin, @Mic92

nand0p commented:
@copumpkin added both functionality of passing entire raw config and/or just override a few default variables. should be ready for merge?

Mic92 commented:
will take a look at it later this day. In case I forget it, feel free to ping me again after 24h.

Mic92 committed:
buildbot: refactor options

- allow to exchange package
- better option types (lists instead of string)
- remove environment variable (was unused)

Mic92 commented:
I tested the module now with "LocalWorker". I made some changes to the module
in this branch: https://github.com/Mic92/nixpkgs/commits/buildbot See commit message for details.

nand0p commented:
@Mic92 cherry-picked your buildbot refactor commit to this PR

Mic92 commented on nixos/modules/services/continuous-integration/buildbot/master.nix:
```diff
@@ -233,19 +229,18 @@ in {
 
       preStart = ''
         mkdir -vp ${cfg.buildbotDir}
         chown -c ${cfg.user}:${cfg.group} ${cfg.buildbotDir}
```
oh, this should have been "-R", or we need fixed uid/gid for buildbot!
